{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "smoking-encounter",
   "metadata": {},
   "source": [
    "## MUSIAL2.0 Benchmark\n",
    "\n",
    "#### `Author: Simon Hackl`\n",
    "#### `Project: The OMPeome of Treponema pallidum`\n",
    "#### `Contact: simon.hackl@uni-tuebingen.de`\n",
    "#### `Date: 15.02.2022`\n",
    "\n",
    "This _Python_ Notebook was used to generate benchmark data and run `MUSIAL2.0` on it with a concomitant runtime and maximum resident set size measurment.\n",
    "\n",
    "_Note: In order to run this notebook `MUSIAL2.0` (https://github.com/Integrative-Transcriptomics/MUSIAL) and `AmberTools TLeap` (https://ambermd.org/AmberTools.php) have to be installed locally. The paths to the two tools can be set in the cell below. The notebook is intended to be run with Windows OS and the WSL enabled._"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "congressional-theorem",
   "metadata": {},
   "source": [
    "### 1. Benchmark Data Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba7c335",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import string\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "tleapPath = \"/mnt/c/users/siha/desktop/tools/AmberTools21/amber20/bin/tleap\"\n",
    "MUSIALpath = \"./MUSIAL-v2.0.jar\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e789bbc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateRandomNucleotideSequence( sequenceLength: int ) :\n",
    "    ''' Generates a random string over the alphabet { A, C, G, T } of the specified length.\n",
    "    In addition global variant positions are assigned with the following probalilities:\n",
    "    Position substitution probability: 1.03*10**-2\n",
    "    Position indel probability: 5.27*10**-4\n",
    "    \n",
    "    :param sequenceLength: The length of the sequence to be generated.\n",
    "    :return: Returns a random nucleotide sequence of the specified length and dictionaries of substitution and indel positions.\n",
    "    '''\n",
    "    nucleotideBases = [ 'A', 'C', 'G', 'T' ]\n",
    "    sequenceString = \"\"\n",
    "    substitutionPositions = { }\n",
    "    indelPositions = { }\n",
    "    for i in range( 0, sequenceLength ) :\n",
    "        referenceBase = nucleotideBases[ random.randint( 0, len( nucleotideBases ) - 1 ) ]\n",
    "        rand = random.random( )\n",
    "        if rand <= ( 5.27 * ( 10 ** -4 ) ) :\n",
    "            # Declare position as indel position.\n",
    "            # Skip if indel occurs in the last 20 positions to avoid out of index variants (with high probability).\n",
    "            if i + 21 > sequenceLength :\n",
    "                continue\n",
    "            else :\n",
    "                indelPositions[ i + 1 ] = referenceBase\n",
    "        elif rand <= ( ( 5.27 * ( 10 ** -4 ) ) + ( 1.03 * ( 10 ** -2 ) ) ) :\n",
    "            # Declare position as substitution position.\n",
    "            substitutionPositions[ i + 1 ] = referenceBase\n",
    "        sequenceString += referenceBase\n",
    "    return sequenceString, substitutionPositions, indelPositions\n",
    "\n",
    "def generateRandomName( randomNames: set = set( ), prefix: str = \"\", length: int = 6 ) :\n",
    "    ''' Generates a random name intended to be used as a sequence or sample name.\n",
    "        To avoid duplicates a set of already used names can be specified.\n",
    "        \n",
    "    :param randomNames: Set of already used names. If not passed no duplication check is conducted.\n",
    "    :param prefix: A prefix to add to the random name.\n",
    "    :param length: The length of the name to be generated.\n",
    "    :return: String of length 6 representing a random sequence name.\n",
    "    '''\n",
    "    getRandomName = lambda : ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(length))\n",
    "    randomName = prefix + getRandomName( )\n",
    "    while randomName in randomNames :\n",
    "        randomName = getRandomName( )\n",
    "    randomNames.add( randomName )\n",
    "    return randomName\n",
    "\n",
    "def generateRandomFasta( sequence: str, sequenceId: str, outdir: str ) :\n",
    "    ''' Writes the passed sequence and sequenceId as .fasta file to the specified outdir.\n",
    "    \n",
    "    :param sequence: The nucleotide sequence to be used.\n",
    "    :param sequenceId: The nucleotide sequences identifier to be used in the fasta header.\n",
    "    :param outdor: The output directory to which the .fasta file is written to.\n",
    "    '''\n",
    "    chunks = [ sequence[ i : i + 80  ] for i in range( 0, len( sequence ), 80 ) ]\n",
    "    if not outdir.endswith( '/' ) :\n",
    "        outdir += '/'\n",
    "    with open( outdir + sequenceId + \".fasta\", \"w\" ) as randomFasta :\n",
    "        randomFasta.write( \">\" + sequenceId + \"\\n\" )\n",
    "        randomFasta.write( \"\\n\".join( chunks ) )\n",
    "        \n",
    "def generateRandomVcfForSequence( sequence: str, substitutionPositions: dict, indelPositions: dict, sequenceId: str, sampleId: str, outdir: str ) :\n",
    "    ''' Writes a random .vcf file yielding variants at the specified substitutionPositions and\n",
    "    indelPositions dictionaries and sequenceId to the specified outdir.\n",
    "    \n",
    "    The specified dictionaries have to map reference sequence positions to reference sequence\n",
    "    nucleotide contents at these positions.\n",
    "    \n",
    "    For each substitution and indel position the .vcf file will have a probability of 0.5 to obtain a variant.\n",
    "    In case of substitutions an non-reference base nucleotide with uniform probability 1/3 is chosen.\n",
    "    In case of indels an additional decision for deletion or insertion, each with a probability of 0.5, is made\n",
    "    and an length k indel is chosen with probability 0.66**(k-1)*0.34.\n",
    "    For insertions an arbitrary nucleotide sequence of the chosen length is inserted.\n",
    "        \n",
    "    The generated .vcf file will be gziped and indexed using tabix.\n",
    "        \n",
    "    :param sequence: The reference sequence.\n",
    "    :param substitutionPositions: Dictionary mapping integers to single-letter strings.\n",
    "                                  Represents positions at which a substitution is allowed wrt. reference.\n",
    "    :param indelPositions: Dictionary mapping integers to single-letter strings.\n",
    "                           Represents positions at which an indel is allowed wrt. reference.\n",
    "    :param sequenceId: The reference sequence identifier.\n",
    "    :param sampleId: The sample identifier.\n",
    "    :param outdir: The directory to which the .vcf file should be written to.\n",
    "    '''\n",
    "    vcfHeader = \"##fileformat=VCFv4.2\\n#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\t\" + sampleId + \"\\n\"\n",
    "    if not outdir.endswith( '/' ) :\n",
    "        outdir += '/'\n",
    "    sortedVariantPositions = sorted( list( substitutionPositions.keys( ) ) + list( indelPositions.keys( ) ) )\n",
    "    with open( outdir + sampleId + \".vcf\", \"w\" ) as sampleFile :\n",
    "        # Write Header\n",
    "        sampleFile.write( vcfHeader )\n",
    "        for pos in sortedVariantPositions :\n",
    "            rand = random.random( )\n",
    "            if rand >= 0.5 :\n",
    "                # Include variant with probability 0.5.\n",
    "                positionString = sequenceId + \"\\t\" + str( pos ) + \"\\t.\\t\"\n",
    "                if pos in substitutionPositions :\n",
    "                    # Assign substitution.\n",
    "                    referenceBase = substitutionPositions[ pos ]\n",
    "                    possibleAlternatives = [ 'A', 'C', 'G', 'T' ]\n",
    "                    possibleAlternatives.remove( referenceBase )\n",
    "                    alternativeBase = possibleAlternatives[ random.randint( 0, len( possibleAlternatives ) - 1 ) ]\n",
    "                    positionString += referenceBase + \"\\t\" + alternativeBase\n",
    "                    positionString += \"\\t100\\t.\\tAC=2;AF=1.00;AN=2;DP=30;MQ=100.00;MQ0=0\\tGT:AD:DP:GQ\\t1/1:0,30:30:100\\n\"\n",
    "                elif pos in indelPositions :\n",
    "                    # Assign indel.\n",
    "                    # Simulate indel length.\n",
    "                    rand = random.random( )\n",
    "                    indelLength = np.random.geometric( 0.35 )\n",
    "                    if rand >= 0.5 :\n",
    "                        # Assign deletion with probability 0.5.\n",
    "                        referenceBase = \"\"\n",
    "                        for delPos in range( pos - 1, pos + indelLength ) :\n",
    "                            referenceBase += sequence[ delPos ]\n",
    "                        alternativeBase = indelPositions[ pos ]\n",
    "                        positionString += referenceBase + \"\\t\" + alternativeBase\n",
    "                        positionString += \"\\t100\\t.\\tAC=2;AF=1.00;AN=2;DP=30;MQ=100.00;MQ0=0\\tGT:AD:DP:GQ\\t1/1:0,30:30:100\\n\"\n",
    "                    else :\n",
    "                        # Assign insertion with probability 0.5.\n",
    "                        referenceBase = indelPositions[ pos ]\n",
    "                        alternativeBase = indelPositions[ pos ]\n",
    "                        for insPos in range( 0, indelLength ) :\n",
    "                            alternativeBase += [ 'A', 'C', 'G', 'T' ][ random.randint( 0, 3 ) ]\n",
    "                        positionString += referenceBase + \"\\t\" + alternativeBase\n",
    "                        positionString += \"\\t100\\t.\\tAC=2;AF=1.00;AN=2;DP=30;MQ=100.00;MQ0=0\\tGT:AD:DP:GQ\\t1/1:0,30:30:100\\n\"\n",
    "                sampleFile.write( positionString )\n",
    "    # Gzip .vcf file.\n",
    "    process = subprocess.Popen(\n",
    "        ['wsl', 'bgzip', outdir + sampleId + \".vcf\" ],\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE\n",
    "    )\n",
    "    stdout, stderr = process.communicate()\n",
    "    # Generate index for .vcf file.\n",
    "    process = subprocess.Popen(\n",
    "        ['wsl', 'tabix', outdir + sampleId + \".vcf.gz\" ],\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE\n",
    "    )\n",
    "    stdout, stderr = process.communicate()\n",
    "    \n",
    "def generateRandomGenesForSequence( sequence: str, sequenceId: str, numGenes: int, outdir: str, generatePdb: bool = False ) :\n",
    "    ''' Writes a genome annotation .gff file for a specified sequence containing numGenes\n",
    "    genes each with a length of 999 bp. In addition a flag can be set to true that induces\n",
    "    the generation of a respective protein structure file in .pdb format using AmberTools tleap\n",
    "        \n",
    "    :param sequence: The reference genome sequence.\n",
    "    :param sequenceId: The reference genome identifier.\n",
    "    :param numGenes: Number of genes to be generated.\n",
    "    :param outdir: The directory to which the output files should be written to.\n",
    "    :param generatePdb: If a .pdb file for the gene should be generated.\n",
    "    '''\n",
    "    gffHeader = \"##sequence-region \" + sequenceId + \" 1 \" +  str( len( sequence ) ) + \"\\n\"\n",
    "    geneNames = set( )\n",
    "    geneStarts = set( )\n",
    "    if not outdir.endswith( '/' ) :\n",
    "        outdir += '/'\n",
    "    with open( outdir + sequenceId + \".gff\", \"w\" ) as gffFile :\n",
    "        # Write Header\n",
    "        gffFile.write( gffHeader )\n",
    "        while numGenes != 0 :\n",
    "            geneStart = random.randint( 1, len( sequence ) )\n",
    "            geneEnd = geneStart + 998\n",
    "            geneIsValid = False\n",
    "            # Check if gene overlaps with any other gene.\n",
    "            while not geneIsValid :\n",
    "                anyInvalid = False\n",
    "                if geneEnd > len( sequence ) :\n",
    "                    anyInvalid = True\n",
    "                else :\n",
    "                    for gs in geneStarts :\n",
    "                        if ( geneStart >= ( gs - 998 ) ) and ( geneStart <= ( gs + 998 ) ) :\n",
    "                            anyInvalid = True\n",
    "                if not anyInvalid :\n",
    "                    geneIsValid = True\n",
    "                    geneName = generateRandomName( geneNames, \"GENE_\", 5 )\n",
    "                    geneNames.add( geneName )\n",
    "                    geneStarts.add( geneStart )\n",
    "                    rand = random.random( )\n",
    "                    if rand <= 0.5 :\n",
    "                        strandedness = \"+\"\n",
    "                    else :\n",
    "                        strandedness = \"-\"\n",
    "                    gffFile.write( sequenceId + \"\\tRefSeq\\tgene\\t\" + str( geneStart ) + \"\\t\" + str( geneEnd ) + \"\\t.\\t\" + strandedness + \"\\t.\\tName=\" + geneName + \"\\n\" )\n",
    "                    if generatePdb :\n",
    "                        generateGenesPdb( sequence[ geneStart - 1 : geneEnd ], geneName, outdir )\n",
    "                else :\n",
    "                    geneStart = random.randint( 1, len( sequence ) )\n",
    "                    geneEnd = geneStart + 998\n",
    "                    geneIsValid = False\n",
    "            numGenes -= 1\n",
    "    return geneNames\n",
    "\n",
    "def generateGenesPdb( geneSequence: str, geneId: str, outdir: str ) :\n",
    "    ''' Writes a dummy .pdb file for a given nucleotide sequence.\n",
    "    \n",
    "    Therefore the genes nucleotide sequence should have a length fully divisible by three.\n",
    "    This sequence is first translated into an amino-acid sequence (without considering stop codons).\n",
    "    A .pdb file, i.e. atom coordinates is than generated by calling AmberTools tleap.\n",
    "    \n",
    "    :param geneSequence: The nucleotide sequence of the gene.\n",
    "    :param geneId: The identifier of the gene, i.e. used for file name.\n",
    "    :param outdir: The directory to which the .pdb file should be written to.\n",
    "    '''\n",
    "    def translate(seq):\n",
    "        table = {\n",
    "            'ATA':'ILE', 'ATC':'ILE', 'ATT':'ILE', 'ATG':'MET',\n",
    "            'ACA':'THR', 'ACC':'THR', 'ACG':'THR', 'ACT':'THR',\n",
    "            'AAC':'ASN', 'AAT':'ASN', 'AAA':'LYS', 'AAG':'LYS',\n",
    "            'AGC':'SER', 'AGT':'SER', 'AGA':'ARG', 'AGG':'ARG',                \n",
    "            'CTA':'LEU', 'CTC':'LEU', 'CTG':'LEU', 'CTT':'LEU',\n",
    "            'CCA':'PRO', 'CCC':'PRO', 'CCG':'PRO', 'CCT':'PRO',\n",
    "            'CAC':'HIS', 'CAT':'HIS', 'CAA':'GLN', 'CAG':'GLN',\n",
    "            'CGA':'ARG', 'CGC':'ARG', 'CGG':'ARG', 'CGT':'ARG',\n",
    "            'GTA':'VAL', 'GTC':'VAL', 'GTG':'VAL', 'GTT':'VAL',\n",
    "            'GCA':'ALA', 'GCC':'ALA', 'GCG':'ALA', 'GCT':'ALA',\n",
    "            'GAC':'ASP', 'GAT':'ASP', 'GAA':'GLU', 'GAG':'GLU',\n",
    "            'GGA':'GLY', 'GGC':'GLY', 'GGG':'GLY', 'GGT':'GLY',\n",
    "            'TCA':'SER', 'TCC':'SER', 'TCG':'SER', 'TCT':'SER',\n",
    "            'TTC':'PHE', 'TTT':'PHE', 'TTA':'LEU', 'TTG':'LEU',\n",
    "            'TAC':'TYR', 'TAT':'TYR', 'TAA':'', 'TAG':'',\n",
    "            'TGC':'CYS', 'TGT':'CYS', 'TGA':'', 'TGG':'TRP',\n",
    "        }\n",
    "        aminoAcids = [ ]\n",
    "        if len( seq ) % 3 == 0 :\n",
    "            for i in range( 0, len( seq ), 3 ) :\n",
    "                codon = seq[ i : i + 3 ]\n",
    "                if table[ codon ] != '' :\n",
    "                    aminoAcids.append( table[ codon ] )\n",
    "        return \" \".join( aminoAcids )\n",
    "    if not outdir.endswith( '/' ) :\n",
    "        outdir += '/'\n",
    "    translatedGeneSequence = translate( geneSequence )\n",
    "    translatedGeneSequenceChunks = [ translatedGeneSequence[ i : i + 400 ] for i in range( 0, len( translatedGeneSequence ), 400 ) ]\n",
    "    tleapSequence = \"{\" + \"\\n\".join( translatedGeneSequenceChunks ) + \"}\\n\"\n",
    "    with open( outdir + geneId + \".tleap\", \"w\", newline=\"\\n\" ) as tleapConf :\n",
    "        tleapConf.writelines( [\n",
    "            \"source oldff/leaprc.ff99\\n\",\n",
    "            \"protein = sequence \" + tleapSequence,\n",
    "            \"savepdb protein \" + outdir + geneId + \".pdb\\n\",\n",
    "            \"quit\\n\", ] )\n",
    "    process = subprocess.Popen(\n",
    "        ['wsl', tleapPath, '-f', outdir + geneId + '.tleap'],\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.PIPE\n",
    "    )\n",
    "    stdout, stderr = process.communicate()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "going-samuel",
   "metadata": {},
   "source": [
    "Running the cell below will start the benchmark data generation. Please ensure to keep the set random seed (cf. comment (2.)) in order to obtain the same benchmark data for each execution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "retained-thailand",
   "metadata": {},
   "outputs": [],
   "source": [
    "# (1.) Create dictionary to store generated data.\n",
    "if not os.path.isdir( \"./D1_TPOMPeome_Hackl2022_BenchmarkData\" ) :\n",
    "    os.mkdir( \"./D1_TPOMPeome_Hackl2022_BenchmarkData\" )\n",
    "# (2.) Set random seed.\n",
    "random.seed( 12345 )\n",
    "np.random.seed( 12345 )\n",
    "# (3.) Simulate 13 genomes with a length ranging from 1Mbp to 13Mbp in 1Mbp steps.\n",
    "genomeNames = set( )\n",
    "for i in range( 1, 14 ) :\n",
    "    genomeSequence, substitutionPositions, indelPositions = generateRandomNucleotideSequence( i * 10**6 )\n",
    "    genomeName = \"GENOME_\" + str ( i ) + \"Mbp\"\n",
    "    genomeNames.add( genomeName )\n",
    "    os.mkdir( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName )\n",
    "    # (3.1) Write simulated genome to .fasta.\n",
    "    generateRandomFasta( genomeSequence, genomeName, \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName )\n",
    "    # (3.2) Generate 300 samples for simulated genome.\n",
    "    sampleNames = set( )\n",
    "    for c in range( 0, 300 ) :\n",
    "        sampleName = generateRandomName( sampleNames, prefix = \"SAMPLE_\", length = 4 )\n",
    "        generateRandomVcfForSequence(\n",
    "            genomeSequence,\n",
    "            substitutionPositions,\n",
    "            indelPositions,\n",
    "            genomeName,\n",
    "            sampleName,\n",
    "            \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName\n",
    "        )\n",
    "    if i == 5 :\n",
    "        # (3.3.1) Simulate 1000 genes and genome annotation for the genome with 5Mbp length.\n",
    "        geneNames = generateRandomGenesForSequence( \n",
    "            genomeSequence,\n",
    "            genomeName,\n",
    "            1000,\n",
    "            \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName,\n",
    "            True\n",
    "        )\n",
    "         # (3.3.2) Prepare MUSIAL gene input configuration .txt file for genome of length 5Mbp.\n",
    "        for genesSize in range( 50, 1050, 50 ) :\n",
    "            with open( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName + \"/\" + \"genes_\" + str( genesSize ) + \".txt\", \"w\", newline=\"\\n\"  ) as gf :\n",
    "                for gn in list(geneNames)[ 0 : genesSize ] :\n",
    "                    gf.write( \",\".join( [ gn, gn, \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName + \"/\" + gn + \".pdb\" ] ) + \"\\n\" )\n",
    "    # (4.) Prepare MUSIAL sample input configuration .txt files.\n",
    "    for sampleSize in range( 30, 330, 30 ) :\n",
    "        with open( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName + \"/\" + \"sample_\" + str( sampleSize ) + \".txt\", \"w\", newline=\"\\n\"  ) as sf :\n",
    "            for sn in list(sampleNames)[ 0 : sampleSize ] :\n",
    "                sf.write( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/\" + genomeName + \"/\" + sn + \".vcf.gz,\" + sn + \"\\n\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "convenient-summit",
   "metadata": {},
   "source": [
    "### 2. Run MUSIAL2.0 on the Dataset\n",
    "The two cells below will start running `MUSIAL2.0` on the generated data and measure the runtime and peak resident set size usage with the `linux time` command."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "interracial-control",
   "metadata": {},
   "source": [
    "#### 2.1 Increasing Sample Set Size and Genome Length, Number of Threads: 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "joined-astronomy",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create files to store results.\n",
    "if not os.path.isdir( \"./R1_TPOMPeome_Hackl2022_Benchmark\" ) :\n",
    "    os.mkdir( \"./R1_TPOMPeome_Hackl2022_Benchmark\" )\n",
    "if not os.path.isfile( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesTimeNT1.txt' ) :\n",
    "    resultsTime = open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesTimeNT1.txt', 'w' )\n",
    "    resultsTime.close( )\n",
    "if not os.path.isfile( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesMemoryNT1.txt' ) :\n",
    "    resultsMem = open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesMemoryNT1.txt', 'w' )\n",
    "    resultsMem.close( )\n",
    "# Iterate over each pair of genome and sample subset and run MUSIAL.\n",
    "for i in range( 1, 14 ) :\n",
    "    genomeName = \"GENOME_\" + str ( i ) + \"Mbp\"\n",
    "    for sampleSize in range( 30, 330, 30 ) :\n",
    "        if not os.path.isdir( \"./temp\" ) :\n",
    "            os.mkdir( \"./temp\" )\n",
    "        try :\n",
    "            process = subprocess.Popen(\n",
    "                [ 'wsl',\n",
    "                  '/usr/bin/time',\n",
    "                  '-v',\n",
    "                  '--output=./D1_TPOMPeome_Hackl2022_BenchmarkData/eval.txt',\n",
    "                  'java',\n",
    "                  '-Xmx30G',\n",
    "                  '-jar', MUSIALpath,\n",
    "                  '-o', 'temp/',\n",
    "                  '-r', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + genomeName + '.fasta',\n",
    "                  '-s', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + 'sample_' + str( sampleSize ) + '.txt'\n",
    "                ],\n",
    "                stdout=subprocess.PIPE,\n",
    "                stderr=subprocess.PIPE\n",
    "            )    \n",
    "            stdout, stderr = process.communicate( )\n",
    "            # Parse and store the measured time and maximum resident set size.\n",
    "            with open( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/eval.txt\", \"r\" ) as evaluation :\n",
    "                line = evaluation.readline( )\n",
    "                while line :\n",
    "                    if line.startswith( \"\\tMaximum resident set size (kbytes):\" ) :\n",
    "                        with open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesMemoryNT1.txt', 'a' ) as resultsMem :\n",
    "                            memoryMb = line.strip( \"\\tMaximum resident set size (kbytes): \" ).strip( )\n",
    "                            memoryMb = int( memoryMb ) / ( 10**3 )\n",
    "                            resultsMem.write(\n",
    "                                \",\".join( [\n",
    "                                    genomeName.strip( \"GENOME_\" ).strip( \"Mbp\" ),\n",
    "                                    str( sampleSize ),\n",
    "                                    str( memoryMb )\n",
    "                                ] ) + \"\\n\"\n",
    "                            )\n",
    "                    if line.startswith( \"\\tElapsed (wall clock) time (h:mm:ss or m:ss):\" ) :\n",
    "                        with open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesTimeNT1.txt', 'a' ) as resultsTime :\n",
    "                            timeS = line.strip( \"\\tElapsed (wall clock) time (h:mm:ss or m:ss): \" ).strip( )\n",
    "                            if timeS.count( \":\" ) == 1 :\n",
    "                                timeS = 60 * int( timeS.split( \":\" )[ 0 ] ) + float( timeS.split( \":\" )[ 1 ] )\n",
    "                            else :\n",
    "                                timeS = 360 * int( timeS.split( \":\" )[ 0 ] ) + 60 * int( timeS.split( \":\" )[ 1 ] ) + float( timeS.split( \":\" )[ 2 ] )\n",
    "                            resultsTime.write(\n",
    "                                \",\".join( [\n",
    "                                    genomeName.strip( \"GENOME_\" ).strip( \"Mbp\" ),\n",
    "                                    str( sampleSize ),\n",
    "                                    str( timeS )\n",
    "                                ] ) + \"\\n\"\n",
    "                            )\n",
    "                    line = evaluation.readline( )\n",
    "        except Exception as e :\n",
    "            continue\n",
    "        shutil.rmtree( \"./temp/\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "democratic-pencil",
   "metadata": {},
   "source": [
    "#### 2.2 Increasing Sample Set and Gene Set Size, Number of Threads: 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "universal-plenty",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create files to store results.\n",
    "if not os.path.isdir( \"./R1_TPOMPeome_Hackl2022_Benchmark\" ) :\n",
    "    os.mkdir( \"./R1_TPOMPeome_Hackl2022_Benchmark\" )\n",
    "if not os.path.isfile( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeTimeNT1.txt' ) :\n",
    "    resultsTime = open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeTimeNT1.txt', 'w' )\n",
    "    resultsTime.close( )\n",
    "if not os.path.isfile( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeMemoryNT1.txt' ) :\n",
    "    resultsMem = open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeMemoryNT1.txt', 'w' )\n",
    "    resultsMem.close( )\n",
    "# Iterate over each pair of genome and sample subset and run MUSIAL.\n",
    "genomeName = \"GENOME_5Mbp\"\n",
    "for genesSize in range( 50, 1050, 50 ) :\n",
    "    for sampleSize in range( 30, 330, 30 ) :\n",
    "        if not os.path.isdir( \"./temp\" ) :\n",
    "            os.mkdir( \"./temp\" )\n",
    "        try :\n",
    "            process = subprocess.Popen(\n",
    "                [ 'wsl',\n",
    "                  '/usr/bin/time',\n",
    "                  '-v',\n",
    "                  '--output=./D1_TPOMPeome_Hackl2022_BenchmarkData/eval.txt',\n",
    "                  'java',\n",
    "                  '-Xmx30G',\n",
    "                  '-jar', MUSIALpath,\n",
    "                  '-o', 'temp/',\n",
    "                  '-r', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + genomeName + '.fasta',\n",
    "                  '-s', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + 'sample_' + str( sampleSize ) + '.txt',\n",
    "                  '-a', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + genomeName + '.gff',\n",
    "                  '-gf', './D1_TPOMPeome_Hackl2022_BenchmarkData/' + genomeName + '/' + 'genes_' + str( genesSize ) + '.txt'\n",
    "                ],\n",
    "                stdout=subprocess.PIPE,\n",
    "                stderr=subprocess.PIPE\n",
    "            )\n",
    "            stdout, stderr = process.communicate( )\n",
    "            with open( \"./D1_TPOMPeome_Hackl2022_BenchmarkData/eval.txt\", \"r\" ) as evaluation :\n",
    "                line = evaluation.readline( )\n",
    "                while line :\n",
    "                    if line.startswith( \"\\tMaximum resident set size (kbytes):\" ) :\n",
    "                        with open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeMemoryNT1.txt', 'a' ) as resultsMem :\n",
    "                            memoryMb = line.strip( \"\\tMaximum resident set size (kbytes): \" ).strip( )\n",
    "                            memoryMb = int( memoryMb ) / ( 10**3 )\n",
    "                            resultsMem.write(\n",
    "                                \",\".join( [\n",
    "                                    str( genesSize ),\n",
    "                                    str( sampleSize ),\n",
    "                                    str( memoryMb )\n",
    "                                ] ) + \"\\n\"\n",
    "                            )\n",
    "                    if line.startswith( \"\\tElapsed (wall clock) time (h:mm:ss or m:ss):\" ) :\n",
    "                        with open( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeTimeNT1.txt', 'a' ) as resultsTime :\n",
    "                            timeS = line.strip( \"\\tElapsed (wall clock) time (h:mm:ss or m:ss): \" ).strip( )\n",
    "                            if timeS.count( \":\" ) == 1 :\n",
    "                                timeS = 60 * int( timeS.split( \":\" )[ 0 ] ) + float( timeS.split( \":\" )[ 1 ] )\n",
    "                            else :\n",
    "                                timeS = 360 * int( timeS.split( \":\" )[ 0 ] ) + 60 * int( timeS.split( \":\" )[ 1 ] ) + float( timeS.split( \":\" )[ 2 ] )\n",
    "                            resultsTime.write(\n",
    "                                \",\".join( [\n",
    "                                    str( genesSize ),\n",
    "                                    str( sampleSize ),\n",
    "                                    str( timeS )\n",
    "                                ] ) + \"\\n\"\n",
    "                            )\n",
    "                    line = evaluation.readline( )\n",
    "        except Exception as e :\n",
    "            continue\n",
    "        shutil.rmtree( \"./temp/\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adequate-complex",
   "metadata": {},
   "source": [
    "### 3. Visualize Results\n",
    "Executing the cell below will generate visualizations for the two benchmark runs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dental-florist",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Imports\n",
    "\"\"\"\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from itertools import product\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import warnings\n",
    "import numpy as np\n",
    "\n",
    "plt.rcParams.update( { \"text.usetex\": False, \"font.family\": \"serif\" } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "clinical-arlington",
   "metadata": {},
   "outputs": [],
   "source": [
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    \n",
    "    # Parse genomesRSS measurement.\n",
    "    dfGenomesRSS = pd.read_csv( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesMemoryNT1.txt', header=None )\n",
    "    dfGenomesRSS.sort_values( by=[0,1], axis=0, inplace=True )\n",
    "    yGenomesRSS = { }\n",
    "    xGenomesRSS = { }\n",
    "    for genomeSize, sampleSize in product( dfGenomesRSS[ 0 ].unique( ), dfGenomesRSS[ 1 ].unique( ) ) :\n",
    "        if genomeSize not in xGenomesRSS :\n",
    "            xGenomesRSS[ genomeSize ] = [ ]\n",
    "            yGenomesRSS[ genomeSize ] = [ ]\n",
    "        xGenomesRSS[ genomeSize ].append( sampleSize ) #genomeSize * 0.010827 * sampleSize\n",
    "        yGenomesRSS[ genomeSize ].append( dfGenomesRSS[ dfGenomesRSS[ 0 ] == genomeSize ][ dfGenomesRSS[ 1 ] == sampleSize ][ 2 ].values[ 0 ] / 1000 )\n",
    "    \n",
    "    # Parse genomesTime measurement.\n",
    "    dfGenomesTime = pd.read_csv( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenomesTimeNT1.txt', header=None )\n",
    "    dfGenomesTime.sort_values( by=[0,1], axis=0, inplace=True )\n",
    "    yGenomesTime = { }\n",
    "    xGenomesTime = { }\n",
    "    for genomeSize, sampleSize in product( dfGenomesTime[ 0 ].unique( ), dfGenomesTime[ 1 ].unique( ) ) :\n",
    "        if genomeSize not in xGenomesTime :\n",
    "            xGenomesTime[ genomeSize ] = [ ]\n",
    "            yGenomesTime[ genomeSize ] = [ ]\n",
    "        xGenomesTime[ genomeSize ].append( sampleSize ) #genomeSize * 0.010827 * sampleSize\n",
    "        yGenomesTime[ genomeSize ].append( dfGenomesTime[ dfGenomesTime[ 0 ] == genomeSize ][ dfGenomesTime[ 1 ] == sampleSize ][ 2 ].values[ 0 ] )\n",
    "    \n",
    "    # Generate plot.\n",
    "    fig, ( ax1, ax2 ) = plt.subplots( nrows = 2, ncols = 1, figsize = ( 13, 13 ), sharex = True )\n",
    "    \n",
    "    # Plot genomesRSS values.\n",
    "    for gS in list( xGenomesRSS.keys( ) ) :\n",
    "        ax1.plot(\n",
    "            xGenomesRSS[ gS ],\n",
    "            yGenomesRSS[ gS ],\n",
    "            color = cm.get_cmap('Spectral')( gS / len( list( xGenomesRSS.keys( ) ) ) ),\n",
    "            marker='o'\n",
    "        )\n",
    "        ax1.set_ylabel( \"Maximal RSS [GB]\", size = 16, labelpad = 10 )\n",
    "        ax1.tick_params( labelsize = 14 )\n",
    "        ax1.set_title( \"Maximal RSS\", size = 18 )\n",
    "        ax1.grid( )\n",
    "\n",
    "    \n",
    "    # Plot genomesTime values.\n",
    "    for gS in list( xGenomesTime.keys( ) ) :\n",
    "        ax2.plot(\n",
    "            xGenomesTime[ gS ],\n",
    "            yGenomesTime[ gS ],\n",
    "            color = cm.get_cmap('Spectral')( gS / len( list( xGenomesRSS.keys( ) ) ) ),\n",
    "            marker='o',\n",
    "            label = \"Genome Size: \" + str( gS ) + \" Mbp\"\n",
    "        )\n",
    "        ax2.set_xlabel( \"Number of Samples\", size = 16, labelpad = 10 )\n",
    "        ax2.set_xticks( np.arange( 30, 330, 30 ) )\n",
    "        ax2.set_ylabel( \"Runtime [s]\", size = 16, labelpad = 10 )\n",
    "        ax2.tick_params( labelsize = 14 )\n",
    "        ax2.set_title( \"Runtime\", size = 18 )\n",
    "        ax2.grid( )\n",
    "    \n",
    "    plt.suptitle( \"Benchmark Results of MUSIAL2.0\\nIncreasing Genome and Sample Set Size\", size = 20, y = 1.01, horizontalalignment='center', verticalalignment='top' )\n",
    "    plt.tight_layout( )\n",
    "    plt.legend( ncol = 4, loc = 'center', bbox_to_anchor = ( 0.5, -0.3 ), fontsize = 14 )\n",
    "    plt.show( )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "headed-commonwealth",
   "metadata": {},
   "outputs": [],
   "source": [
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    \n",
    "    # Parse genesRSS measurement.\n",
    "    dfGenesRSS = pd.read_csv( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeMemoryNT1.txt', header=None )\n",
    "    dfGenesRSS.sort_values( by=[0,1], axis=0, inplace=True )\n",
    "    yGenesRSS = { }\n",
    "    xGenesRSS = { }\n",
    "    for genesSize, sampleSize in product( dfGenesRSS[ 0 ].unique( ), dfGenesRSS[ 1 ].unique( ) ) :\n",
    "        if genesSize % 100 == 0 and not genesSize in xGenesRSS :\n",
    "            xGenesRSS[ genesSize ] = [ ]\n",
    "            yGenesRSS[ genesSize ] = [ ]\n",
    "        if genesSize % 100 == 0 :\n",
    "            xGenesRSS[ genesSize ].append( sampleSize ) #genomeSize * 0.010827 * sampleSize\n",
    "            yGenesRSS[ genesSize ].append( dfGenesRSS[ dfGenesRSS[ 0 ] == genesSize ][ dfGenesRSS[ 1 ] == sampleSize ][ 2 ].values[ 0 ] / 1000 )\n",
    "    \n",
    "    # Parse genesTime measurement.\n",
    "    dfGenesTime = pd.read_csv( './R1_TPOMPeome_Hackl2022_Benchmark/IncSamplesAndGenes5MbpGenomeTimeNT1.txt', header=None )\n",
    "    dfGenesTime.sort_values( by=[0,1], axis=0, inplace=True )\n",
    "    yGenesTime = { }\n",
    "    xGenesTime = { }\n",
    "    for genesSize, sampleSize in product( dfGenesTime[ 0 ].unique( ), dfGenesTime[ 1 ].unique( ) ) :\n",
    "        if genesSize % 100 == 0 and not genesSize in xGenesTime :\n",
    "            xGenesTime[ genesSize ] = [ ]\n",
    "            yGenesTime[ genesSize ] = [ ]\n",
    "        if genesSize % 100 == 0 :\n",
    "            xGenesTime[ genesSize ].append( sampleSize ) #genomeSize * 0.010827 * sampleSize\n",
    "            yGenesTime[ genesSize ].append( dfGenesTime[ dfGenesTime[ 0 ] == genesSize ][ dfGenesTime[ 1 ] == sampleSize ][ 2 ].values[ 0 ] )\n",
    "        \n",
    "    # Generate plot.\n",
    "    fig, ( ax1, ax2 ) = plt.subplots( nrows = 2, ncols = 1, figsize = ( 13, 13 ), sharex = True )\n",
    "    \n",
    "    # Plot genesRSS values.\n",
    "    for gS in list( xGenesRSS.keys( ) ) :\n",
    "        ax1.plot(\n",
    "            xGenesRSS[ gS ],\n",
    "            yGenesRSS[ gS ],\n",
    "            color = cm.get_cmap('Spectral')( gS / 1000 ),\n",
    "            marker='o'\n",
    "        )\n",
    "        ax1.set_ylabel( \"Maximal RSS [GB]\", size = 16, labelpad = 10 )\n",
    "        ax1.tick_params( labelsize = 14 )\n",
    "        ax1.set_title( \"Maximal RSS\", size = 18 )\n",
    "        ax1.grid( b = True )\n",
    "\n",
    "    \n",
    "    # Plot genesTime values.\n",
    "    for gS in list( xGenesTime.keys( ) ) :\n",
    "        ax2.plot(\n",
    "            xGenesTime[ gS ],\n",
    "            yGenesTime[ gS ],\n",
    "            color = cm.get_cmap('Spectral')( gS / 1000 ),\n",
    "            marker='o',\n",
    "            label = \"Gene Set Size: \" + str( gS )\n",
    "        )\n",
    "        ax2.set_xlabel( \"Number of Samples\", size = 16, labelpad = 10 )\n",
    "        ax2.set_xticks( np.arange( 30, 330, 30 ) )\n",
    "        ax2.set_ylabel( \"Runtime [s]\", size = 16, labelpad = 10 )\n",
    "        ax2.tick_params( labelsize = 14 )\n",
    "        ax2.set_title( \"Runtime\", size = 18 )\n",
    "        ax2.grid( b = True )\n",
    "    \n",
    "    plt.suptitle( \"Benchmark Results of MUSIAL2.0\\nIncreasing Gene Set and Sample Set Size, Genome Size 5 Mbp\", size = 20, y = 1.01, horizontalalignment='center', verticalalignment='top' )\n",
    "    plt.tight_layout( )\n",
    "    plt.legend( ncol = 4, loc = 'center', bbox_to_anchor = ( 0.5, -0.3 ), fontsize = 14 )\n",
    "    plt.show( )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
