{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "concrete-toner",
   "metadata": {},
   "source": [
    "## Protein Structure Preparation\n",
    "\n",
    "#### `Author: Simon Hackl`\n",
    "#### `Project: The OMPeome of Treponema pallidum`\n",
    "#### `Contact: simon.hackl@uni-tuebingen.de`\n",
    "#### `Date: 15.02.2022`\n",
    "\n",
    "This _Python_ Notebook guides through and documents the steps of protein structure preparation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58758ced",
   "metadata": {},
   "source": [
    "### 1. Collect Protein Structures\n",
    "For this project a set of $36$ OMP coding genes of interest were considered. For $24$ of them a structure was already resolved by Hawley _et al._ (https://doi.org/10.1128/JB.00082-21). The respective structures were downloaded from the publication and stored in the `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ProteinStructures/` directory.\n",
    "\n",
    "For the remaining $12$ OMPs no structure was known and thus the following procedure was applied to predict respective structures."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02c5add8",
   "metadata": {},
   "source": [
    "### 2. Extraction and Translation of Gene Sequences\n",
    "\n",
    "Starting point for the protein structure prediction was the extraction of their respective genes nucleotide sequences. The chosen reference genome sequence and annotation were downloaded from https://www.ncbi.nlm.nih.gov/nuccore/NC_021490.2 and stored at `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ReferenceGenome/NC_021490.2.fasta` and `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ReferenceGenome/NC_021490.2.gff3`. A list with the names of the OMP coding genes of interest with no resolved structure was created at `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureGeneNames.txt`.\n",
    "\n",
    "The respective locus and strand orientation of the genes in the chosen reference was detected by matching a TPXXXX gene name with the old_locus_tag attributes value (format: TPANIC\\_XXXX) from the reference genome annotation.\n",
    "\n",
    "Next, the genes nucleotide sequences were extracted from the reference genome acoording to this information and translated into their repsective amino-acid sequences using `Expasy Translate` tool (https://web.expasy.org/translate/). The obtained raw sequences were stored at `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureSequences.fasta`. The sequences were truncated to their maximal length open reading frame (indicated by `Expasy Translate`), i.e. if the sequence did not start with an M residue, all residues before the first M were truncated. Affected sequences were appended the suffix \\_Morf to their header. The resulting sequences were stored at `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureSequences_MaxORF.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9d7bb9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from subprocess import Popen, PIPE\n",
    "\n",
    "# Parse information from the genome annotation and store it in a dictionary accessible by the old_locus_tag attribute.\n",
    "annotatedGenes = { }\n",
    "with open( \"./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ReferenceGenome/NC_021490.2.gff3\", \"r\" ) as annotatedGenesFile :\n",
    "    line = annotatedGenesFile.readline( )\n",
    "    while line :\n",
    "        if line.startswith( \"##\" ) : # Skip comment lines\n",
    "            line = annotatedGenesFile.readline( )\n",
    "            continue\n",
    "        else :\n",
    "            splitLine = line.split( \"\\t\" )\n",
    "            if splitLine[ 0 ] == \"NC_021490.2\" and splitLine[ 1 ] == \"RefSeq\" and splitLine[ 2 ] == \"gene\" :\n",
    "                attributes = splitLine[ 8 ].strip( ).split( \";\" )\n",
    "                for attribute in attributes :\n",
    "                    key, value = attribute.split( \"=\" )\n",
    "                    if key == \"old_locus_tag\" :\n",
    "                        annotatedGenes[ value ] = {\n",
    "                            \"locusStart\": int( splitLine[ 3 ] ),\n",
    "                            \"locusEnd\": int( splitLine[ 4 ] ),\n",
    "                            \"strandOrientation\": splitLine[ 6 ]\n",
    "                        }\n",
    "        line = annotatedGenesFile.readline( )\n",
    "        \n",
    "# Parse the reference genome sequence.\n",
    "referenceGenome = \"\"\n",
    "with open( \"./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ReferenceGenome/NC_021490.2.fasta\", \"r\" ) as referenceGenomeFile :\n",
    "    line = referenceGenomeFile.readline( ) # Skip fasta header\n",
    "    line = referenceGenomeFile.readline( )\n",
    "    while line :\n",
    "        referenceGenome += line.strip( )\n",
    "        line = referenceGenomeFile.readline( )\n",
    "        \n",
    "# For each of the OMP coding genes of interest the respective gene sequence is extracted and translated by accessing\n",
    "# the Expasy Translate tool.\n",
    "with open( \"./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureGeneNames.txt\", \"r\" ) as OMPGenesFile :\n",
    "    translatedGenes = { }\n",
    "    line = OMPGenesFile.readline( )\n",
    "    while line :\n",
    "        geneName = line.strip( )\n",
    "        print( \"Processing Gene: \" + geneName )\n",
    "        geneOldLocusTag = \"TPANIC_\" + geneName[ 2: ]\n",
    "        geneStart = annotatedGenes[ geneOldLocusTag ][ \"locusStart\" ]\n",
    "        geneEnd = annotatedGenes[ geneOldLocusTag ][ \"locusEnd\" ]\n",
    "        geneOrientation = annotatedGenes[ geneOldLocusTag ][ \"strandOrientation\" ]\n",
    "        geneSequence = referenceGenome[ geneStart - 1 : geneEnd ]\n",
    "        print( \"- %Ns: \" + str( geneSequence.count( 'N' ) / len( geneSequence ) ) )\n",
    "        print( \"- Orientation: \" + geneOrientation )\n",
    "        process = Popen(\n",
    "            [\n",
    "                'wsl', \n",
    "                'curl',\n",
    "                '-s',\n",
    "                '-d',\n",
    "                \"'dna_sequence='\" + geneSequence + \"'&output_format=fasta'\",\n",
    "                'https://web.expasy.org/cgi-bin/translate/dna2aa.cgi'\n",
    "                #'>',\n",
    "                #'./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/OMPSequences/' + geneName + '.fasta'\n",
    "            ],\n",
    "            stdout=PIPE, stderr=PIPE\n",
    "        )\n",
    "        stdout, stderr = process.communicate( )\n",
    "        response = stdout.decode( encoding = 'UTF-8' ).split( \"\\n\" )\n",
    "        translatedSequence = \"\"\n",
    "        entryCounter = 0\n",
    "        for entry in response :\n",
    "            if entry.startswith( \">\" ) :\n",
    "                entryCounter += 1\n",
    "                continue\n",
    "            if geneOrientation == \"-\" and entryCounter == 1 :\n",
    "                translatedSequence += entry\n",
    "            elif geneOrientation == \"-\" and entryCounter > 1 :\n",
    "                break\n",
    "            elif geneOrientation == \"+\" and entryCounter == 4 :\n",
    "                translatedSequence += entry\n",
    "            elif geneOrientation == \"+\" and entryCounter > 4 :\n",
    "                break\n",
    "        print( \"- Translated Sequence: \\n\" + translatedSequence.strip( \"-\" ) )\n",
    "        translatedGenes[ geneName ] = translatedSequence.strip( \"-\" )\n",
    "        line = OMPGenesFile.readline( )\n",
    "        print( )\n",
    "with open( \"./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureSequences.fasta\", \"w+\" ) as OMPSequencesFile :\n",
    "    for key, value in translatedGenes.items( ) :\n",
    "        OMPSequencesFile.write( \">\" + key + \"\\n\" )\n",
    "        sequenceChunks = [ value[ i : i + 80 ] for i in range( 0, len( value ), 80 ) ]\n",
    "        for sequenceChunk in sequenceChunks :\n",
    "            OMPSequencesFile.write( sequenceChunk + \"\\n\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "267a8a5f",
   "metadata": {},
   "source": [
    "### 3. Truncation Signal Peptides\n",
    "\n",
    "In the next step possible signal peptides were remove from the translated genes sequences in advance before structure prediction. To do so, the file `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureSequences_MaxORF.fasta` was used as input for `SignalP 5.0` (https://services.healthtech.dtu.dk/service.php?SignalP-5.0) with the following parameters:\n",
    "\n",
    "    - Output format: Long output\n",
    "    - Organism group: Gram-negative\n",
    "    \n",
    "The results of the `SignalP 5.0` signal peptide prediction were stored at  `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/SignalP_Results.zip` and the amino-acid sequences with the respective predicted signal peptides being removed (this process was conducted manually, the suffix \\_trc was added to the respective fasta headers if any predicted signal peptide was removed) were stored at  `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/TPOMPNoStructureSequences_MaxORF_SignalP.fasta`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a25ee5",
   "metadata": {},
   "source": [
    "### 4. Structure Prediction Using Robetta/RoseTTAFold \n",
    "\n",
    "The protein sequences were finally used as input sequences for structure prediction using `RoseTTAFold` via the `Robetta` web server (https://robetta.bakerlab.org/). The predicted structures were stored at `./R3_TPOMPeome_Hackl2022_ProteinStructurePreparation/ProteinStructures/`. The file names will include\n",
    "    \n",
    "    - .morf If some residues were truncated from the start of the sequence in order to obtain the max. open reading frame.\n",
    "    - .trc If some signal peptide sequence was truncated.\n",
    "    - .pred If they were predicted herein."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
