{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "concrete-toner",
   "metadata": {},
   "source": [
    "## Variant Calling\n",
    "\n",
    "#### `Author: Simon Hackl`\n",
    "#### `Project: The OMPeome of Treponema pallidum`\n",
    "#### `Contact: simon.hackl@uni-tuebingen.de`\n",
    "#### `Date: 15.02.2022`\n",
    "\n",
    "This _Python_ Notebook guides through and documents the steps of variant calling."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7672aec7",
   "metadata": {},
   "source": [
    "### 1. Acquisition of Reference Genome and Sample Data\n",
    "The chosen reference genome sequence and annotation were downloaded from https://www.ncbi.nlm.nih.gov/nuccore/NC_021490.2 and stored at `./R4_TPOMPeome_Hackl2022_VariantCalling/ReferenceGenome/NC_021490.2.fasta` and `./R4_TPOMPeome_Hackl2022_VariantCalling/ReferenceGenome/NC_021490.2.gff3`.\n",
    "\n",
    "A set of $74$ DNA sequencing read datasets already processed with `Clip&Merge` were collected from the _romanov_ server of the _Integrative Transcriptomics_ group and stored in the `./R4_TPOMPeome_Hackl2022_VariantCalling/SampleReads/` directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "selective-accreditation",
   "metadata": {},
   "source": [
    "### 2. Variant Calling\n",
    "\n",
    "Genotyping was conducted in three modes for each sample by invoking the bash script below in the directory `./R4_TPOMPeome_Hackl2022_VariantCalling/`:\n",
    "    \n",
    "    1. Genotyping with `GATK UnifiedGenotyper` without RecoverMI applied ().\n",
    "       Filename: <Samplename>-NC_021490.2-UG-variants.vcf\n",
    "       \n",
    "    2. Genotyping with `GATK UnifiedGenotyper` with RecoverMI applied.\n",
    "       Filename: <Samplename>-NC_021490.2-UG-variants.fxd.vcf\n",
    "       \n",
    "    3. Genotyping with `GATK HaplotypeCaller` with RecoverMI applied.\n",
    "       Filename: <Samplename>-NC_021490.2-HC-variants.fxd.vcf\n",
    "    \n",
    "In advance, `DeDup` was applied to true sample reads, i.e. those reads that were derived from genomes were not treated. The resulting vcf files were stored at `./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling` with the file names as explained above.\n",
    "\n",
    "**NOTE:** In order to run the script, `bwa`, `samtools`, `GATK` and `DeDup` have to be installed and accessible via the path variables. `RecoverMI.py` has to be stored in the `./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling` directory.\n",
    "\n",
    "```\n",
    "#!/bin/bash\n",
    "\n",
    "# Create output directory.\n",
    "mkdir ./VariantCalling/\n",
    "\n",
    "# Generate reference genome indices.\n",
    "bwa index ./ReferenceGenome/NC_021490.2.fasta\n",
    "samtools faidx ./ReferenceGenome/NC_021490.2.fasta\n",
    "samtools dict ./ReferenceGenome/NC_021490.2.fasta > ./ReferenceGenome/NC_021490.2.dict\n",
    "\n",
    "# Run deduplication, multi-map preservation and variant calling per sample.\n",
    "for sampleDir in ./SampleReads/*\n",
    "do\n",
    "\tsampleName=$(basename \"${sampleDir}\")\n",
    "\tif [ ! -d \"./VariantCalling/${sampleName}/\" ]; then\n",
    "\t\tmkdir ./VariantCalling/${sampleName}\n",
    "\t\techo ${sampleName}\n",
    "\t\tfor sampleFASTQFile in ./SampleReads/$sampleName/*.fq.gz\n",
    "\t\tdo\n",
    "\t\t\tbwa mem ./ReferenceGenome/NC_021490.2.fasta ${sampleFASTQFile} -t 8 -a -R '@RG\\tID:1\\tPL:ILLUMINA\\tLB:lib1\\tSM:A2_1' | samtools sort -@ 8 -o ./VariantCalling/${sampleName}/mapped.bam\n",
    "\t\t\t# If samples are reads derived from genomes do not apply DeDup.\n",
    "\t\t\tif [[ ${sampleName} == \"BosniaA\" || ${sampleName} == \"CDC2\" || ${sampleName} == \"Chicago\" || ${sampleName} == \"Dallas\" || ${sampleName} == \"Fribourg\" || ${sampleName} == \"Gauthier\" || ${sampleName} == \"MexicoA\" || ${sampleName} == \"Nichols\" || ${sampleName} == \"SamoaD\" || ${sampleName} == \"Seattle81\" || ${sampleName} == \"SS14\" || ${sampleName} == \"Cuniculi\" ]]; then\n",
    "\t\t\t\tsamtools index -@ 8 ./VariantCalling/${sampleName}/mapped.bam\n",
    "\t\t\t\t# Run variant valling with UnifiedGenotyper on mapping without preserved multi-mappings.\n",
    "\t\t\t\tgatk -T UnifiedGenotyper -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.bam -nt 4 -out_mode EMIT_ALL_SITES -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-UG-variants.vcf\n",
    "\n",
    "\t\t\t\t# Sort by read name and apply RecoverMI.\n",
    "\t\t\t\tsamtools sort -n -@ 8 -o ./VariantCalling/${sampleName}/mapped.sorted.sam ./VariantCalling/${sampleName}/mapped.bam\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.bam\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.bam.*\n",
    "\t\t\t\t./RecoverMI.py ./VariantCalling/${sampleName}/mapped.sorted.sam ./VariantCalling/${sampleName}/\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.sorted.sam\n",
    "\t\t\t\t# Convert sam to bam, re-sort by position and index bam file.\n",
    "\t\t\t\tsamtools sort -o ./VariantCalling/${sampleName}/mapped.sorted.fxd.bam -@ 8 ./VariantCalling/${sampleName}/mapped.sorted.fxd.sam\n",
    "\t\t\t\tsamtools index -@ 8 ./VariantCalling/${sampleName}/mapped.sorted.fxd.bam\n",
    "\n",
    "\t\t\t\t# Run variant calling with HaplotypeCaller on mapping with preserved multi-mappings.\n",
    "\t\t\t\tgatk -T HaplotypeCaller -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.sorted.fxd.bam -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-HC-variants.fxd.vcf\n",
    "\t\t\t\t# Run variant valling with UnifiedGenotyper on mapping with preserved multi-mappings.\n",
    "\t\t\t\tgatk -T UnifiedGenotyper -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.sorted.fxd.bam -nt 4 -out_mode EMIT_ALL_SITES -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-UG-variants.fxd.vcf \n",
    "\t\t\telse\n",
    "\t\t\t\t# Run DeDup to remove PCR duplicates.\n",
    "\t\t\t\tdedup -i ./VariantCalling/${sampleName}/mapped.bam -o ./VariantCalling/${sampleName}/\n",
    "\t\t\t\tmv ./VariantCalling/${sampleName}/mapped_rmdup.bam ./VariantCalling/${sampleName}/mapped.rmdup.bam\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.bam\n",
    "\t\t\n",
    "\t\t\t\tsamtools sort -@ 4 -o ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam ./VariantCalling/${sampleName}/mapped.rmdup.bam\n",
    "\t\t\t\tsamtools index -@ 4 ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam\n",
    "\t\t\t\t# Run variant valling with UnifiedGenotyper on mapping without preserved multi-mappings.\n",
    "\t\t\t\tgatk -T UnifiedGenotyper -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam -nt 4 -out_mode EMIT_ALL_SITES -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-UG-variants.vcf\n",
    "\n",
    "\t\t\t\t# Sort by read name and apply RecoverMI.\n",
    "\t\t\t\tsamtools sort -@ 4 -n -o ./VariantCalling/${sampleName}/mapped.rmdup.sortedNames.sam ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.rmdup.sorted.bam.*\n",
    "\t\t\t\t./Scripts/RecoverMI.py ./VariantCalling/${sampleName}/mapped.rmdup.sortedNames.sam ./VariantCalling/${sampleName}/\n",
    "\t\t\t\trm ./VariantCalling/${sampleName}/mapped.rmdup.sortedNames.sam\n",
    "\t\t\t\t## Convert sam to bam, re-sort by position and index bam file.\n",
    "\t\t\t\tsamtools sort -o ./VariantCalling/${sampleName}/mapped.rmdup.sorted.fxd.bam -@ 4 ./VariantCalling/${sampleName}/mapped.rmdup.sortedNames.fxd.sam\n",
    "\t\t\t\tsamtools index -@ 4 ./VariantCalling/${sampleName}/mapped.rmdup.sorted.fxd.bam\n",
    "\n",
    "\t\t\t\t# Run variant calling with HaplotypeCaller on mapping with preserved multi-mappings.\n",
    "\t\t\t\tgatk -T HaplotypeCaller -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.rmdup.sorted.fxd.bam -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-HC-variants.fxd.vcf\n",
    "\t\t\t\t# Run variant valling with UnifiedGenotyper on mapping with preserved multi-mappings.\n",
    "\t\t\t\tgatk -T UnifiedGenotyper -R ./ReferenceGenome/NC_021490.2.fasta -I ./VariantCalling/${sampleName}/mapped.rmdup.sorted.fxd.bam -nt 4 -out_mode EMIT_ALL_SITES -o ./VariantCalling/${sampleName}/${sampleName}-NC_021490.2-UG-variants.fxd.vcf \n",
    "\t\t\tfi\n",
    "\t\tbreak\n",
    "\t\tdone\n",
    "\tfi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31a89502",
   "metadata": {},
   "source": [
    "### 3. Assessing Repeat Resolution\n",
    "\n",
    "In order to assess if `RecoverMI` worked correctly, the respective vcf files generated by `GATK UnifiedGenotyper` with and without prior application of `RecoverMI` were compared by calculating (wrt. gene length) the percentage of positions that were disregarded and actually located in a repeat.\n",
    "\n",
    "For this, repeats were first identified using `DACCOR`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "170c6cfd",
   "metadata": {},
   "source": [
    "### 3.1 Identification of Min. Read Length\n",
    "In order to choose appropriate parameters for `DACCOR` the minimum read length of the dataset was extracted from the respective FASTQ files by using the bash script below (in the `./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling` directory).\n",
    "\n",
    "```\n",
    "#!/bin/sh\n",
    "for readFile in ./SampleReads/*/*.fq.gz\n",
    "do\n",
    "\techo \"Processing: ${readFile}\"\n",
    "\tzcat ${readFile} | awk '{if(NR%4==2) print length($1)}' | sort -n | uniq -c >> ./rl.txt\n",
    "done\n",
    "sort -n ./rl.txt | uniq -c >> ./UniqueReadLengths.txt\n",
    "rm ./rl.txt\n",
    "```\n",
    "\n",
    "From the `UniqueReadLengths.txt` file the shortest read length was determined to be $30$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68976f57",
   "metadata": {},
   "source": [
    "### 3.2 Extraction of Gene Sequences\n",
    "\n",
    " A list with the names of the OMP coding genes of interest was created at `./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGeneNames.txt`.\n",
    "\n",
    "The respective locus and strand orientation of the genes in the chosen reference was detected by matching a TPXXXX gene name with the old_locus_tag attributes value (format TPANIC\\_XXXX) from the reference genome annotation.\n",
    "\n",
    "Next, the genes nucleotide sequences were extracted from the reference genome acoording to this information. The obtained sequences were stored at `./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGenesSequences/`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e09b1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import json\n",
    "\n",
    "from subprocess import Popen, PIPE\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pylab as plt\n",
    "\n",
    "plt.rcParams.update( { \"text.usetex\": False, \"font.family\": \"serif\" } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a6ab16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse information from the genome annotation and store it in a dictionary accessible by the old_locus_tag attribute.\n",
    "annotatedGenes = { }\n",
    "with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/ReferenceGenome/NC_021490.2.gff3\", \"r\" ) as annotatedGenesFile :\n",
    "    line = annotatedGenesFile.readline( )\n",
    "    while line :\n",
    "        if line.startswith( \"##\" ) : # Skip comment lines\n",
    "            line = annotatedGenesFile.readline( )\n",
    "            continue\n",
    "        else :\n",
    "            splitLine = line.split( \"\\t\" )\n",
    "            if splitLine[ 0 ] == \"NC_021490.2\" and splitLine[ 1 ] == \"RefSeq\" and ( splitLine[ 2 ] == \"gene\" or splitLine[ 2 ] == \"pseudogene\" )  :\n",
    "                attributes = splitLine[ 8 ].strip( ).split( \";\" )\n",
    "                for attribute in attributes :\n",
    "                    key, value = attribute.split( \"=\" )\n",
    "                    if key == \"old_locus_tag\" :\n",
    "                        annotatedGenes[ value ] = {\n",
    "                            \"locusStart\": int( splitLine[ 3 ] ),\n",
    "                            \"locusEnd\": int( splitLine[ 4 ] ),\n",
    "                            \"strandOrientation\": splitLine[ 6 ]\n",
    "                        }\n",
    "        line = annotatedGenesFile.readline( )\n",
    "        \n",
    "# Parse the reference genome sequence.\n",
    "referenceGenome = \"\"\n",
    "with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/ReferenceGenome/NC_021490.2.fasta\", \"r\" ) as referenceGenomeFile :\n",
    "    line = referenceGenomeFile.readline( ) # Skip fasta header\n",
    "    line = referenceGenomeFile.readline( )\n",
    "    while line :\n",
    "        referenceGenome += line.strip( )\n",
    "        line = referenceGenomeFile.readline( )\n",
    "        \n",
    "# For each of the OMP coding genes of interest the respective gene sequence is extracted.\n",
    "if not os.path.isdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGenesSequences/\" ) :\n",
    "    os.mkdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGenesSequences/\" )\n",
    "with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGeneNames.txt\", \"r\" ) as OMPGenesFile :\n",
    "    line = OMPGenesFile.readline( )\n",
    "    while line :\n",
    "        geneName = line.strip( )\n",
    "        print( \"Processing Gene: \" + geneName )\n",
    "        geneOldLocusTag = \"TPANIC_\" + geneName[ 2: ]\n",
    "        geneStart = annotatedGenes[ geneOldLocusTag ][ \"locusStart\" ]\n",
    "        geneEnd = annotatedGenes[ geneOldLocusTag ][ \"locusEnd\" ]\n",
    "        geneOrientation = annotatedGenes[ geneOldLocusTag ][ \"strandOrientation\" ]\n",
    "        geneSequence = referenceGenome[ geneStart - 1 : geneEnd ]\n",
    "        with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGenesSequences/\" + geneName + \".fasta\", \"w+\" ) as OMPGeneSequenceFile :\n",
    "            OMPGeneSequenceFile.write( \">\" + geneName + \"\\n\" )\n",
    "            sequenceChunks = [ geneSequence[ i : i + 80 ] for i in range( 0, len( geneSequence ), 80 ) ]\n",
    "            for sequenceChunk in sequenceChunks :\n",
    "                OMPGeneSequenceFile.write( sequenceChunk + \"\\n\" )\n",
    "        line = OMPGenesFile.readline( )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5dacdff",
   "metadata": {},
   "source": [
    "### 3.3 Identify Repeats with DACCOR identify\n",
    "\n",
    "To identify repeats with `DACCOR` on gene and genome level the following bash script was run in `./R4_TPOMPeome_Hackl2022_VariantCalling/`.\n",
    "\n",
    "**NOTE:** In order to run the script, `DACCOR` has to be installed and accessible via the path variables.\n",
    "\n",
    "```\n",
    "#!/bin/sh\n",
    "mkdir ./DACCORIdentifyResults/\n",
    "echo '-- Run DACCOR identify on single OMP genes'\n",
    "FILES=\"./TPOMPGenesSequencess/TP*.fasta\"\n",
    "for file in $FILES\n",
    "do\n",
    "\techo \"- Processing file: $file\"\n",
    "\tfileBasename=`basename ${file%.fasta}`\n",
    "\tmkdir ./DACCORIdentifyResults/${fileBasename}\n",
    "\toutDir=./DACCORIdentifyResults/${fileBasename}\n",
    "\tdaccor identify -i ${file} -o ${outDir} -m 0 -M 0 -rl 30 -k 15 -ws\n",
    "done\n",
    "\n",
    "echo '-- Run identify subprogram on full genome NC_021490.2'\n",
    "file=\"./ReferenceGenome/NC_021490.2.fasta\"\n",
    "outDir=./DACCORIdentifyResults/NC_021490.2\n",
    "daccor identify -i ${file} -o ${outDir} -m 0 -M 0 -rl 30 -k 15 -ws'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aed32487",
   "metadata": {},
   "source": [
    "### 3.4 Identify Repeats Interfering with Gene Loci\n",
    "Next, in order to treat only repeats that truely interfer with gene positions the OMP genes positions are compared with the repeat positions and, for each gene, the percentage of gene positions that are actual repeat positions is computed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "black-warren",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneNames = list( )\n",
    "with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/TPOMPGeneNames.txt\", \"r\" ) as OMPGenesFile :\n",
    "    line = OMPGenesFile.readline( )\n",
    "    while line :\n",
    "        geneNames.append( line.strip( ) )\n",
    "        line = OMPGenesFile.readline( )\n",
    "        \n",
    "# Parse identified repeats.\n",
    "repeats = { }\n",
    "for directory in os.listdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/DACCORIdentifyResults/\" ) :\n",
    "    for file in os.listdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/DACCORIdentifyResults/\" + directory + \"/\" ) :\n",
    "        if file.endswith( \"_Summary.csv\" ) :\n",
    "            isRepeatInfo = False\n",
    "            with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/DACCORIdentifyResults/\" + directory + \"/\" + file, \"r\" ) as repeatsSummary :\n",
    "                line = True\n",
    "                isRepeatInfo = False\n",
    "                while line :\n",
    "                    line = repeatsSummary.readline( )\n",
    "                    if line.startswith( \"fastaName\\t\" ) :\n",
    "                        isRepeatInfo = True\n",
    "                        continue\n",
    "                    if isRepeatInfo and len( line.split( \"\\t\") ) > 1 :\n",
    "                        offset = 0\n",
    "                        if directory.startswith( \"TP\" ) :\n",
    "                            offset = annotatedGenes[ \"TPANIC_\" + directory[ 2: ] ][ 'locusStart' ] - 1\n",
    "                        if line.startswith( \"NC_\" ) : # Treat _ in reference name with index shift.\n",
    "                            repeats[ directory + \"-\" + line.split( \"\\t\" )[ 1 ] ] = {\n",
    "                                \"length\": int( line.split( \"\\t\" )[ 0 ].split( \"_\" )[ 5 ] ),\n",
    "                                \"starts\": [ int( s ) for s in line.split( \"\\t\" )[ 0 ].split( \"_\" )[ 3 ].strip( \"]\" ).strip( \"[\" ).split( \",\" ) ]\n",
    "                            }\n",
    "                        else :\n",
    "                            repeats[ directory + \"-\" + line.split( \"\\t\" )[ 1 ] ] = {\n",
    "                                \"length\": int( line.split( \"\\t\" )[ 0 ].split( \"_\" )[4] ),\n",
    "                                \"starts\": [ int( s ) + offset for s in line.split( \"\\t\" )[ 0 ].split( \"_\" )[2].strip( \"]\" ).strip( \"[\" ).split( \",\" ) ]\n",
    "                            }\n",
    "                            \n",
    "relevantRepeats = { }\n",
    "affectedGenes = { }\n",
    "for geneName in geneNames :\n",
    "    geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "    genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "    geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "    totalPositions = 0\n",
    "    for repeatName, repeatInfo in repeats.items( ) :\n",
    "        commonPositions = 0\n",
    "        for repeatStart in repeatInfo[ 'starts' ] :\n",
    "            repeatPositions = list( range( repeatStart, repeatStart + repeatInfo[ 'length' ] - 1 ) )\n",
    "            for gp in genePositions :\n",
    "                for rp in repeatPositions :\n",
    "                    if gp == rp :\n",
    "                        commonPositions += 1\n",
    "                        relevantRepeats[ repeatName ] = repeats[ repeatName ]\n",
    "        totalPositions += commonPositions\n",
    "        if commonPositions > 0 :\n",
    "            if not geneName in affectedGenes :\n",
    "                affectedGenes[ geneName ] = { }\n",
    "            if not repeatName in affectedGenes[ geneName ] :\n",
    "                affectedGenes[ geneName ][ repeatName ] = [ ]\n",
    "            affectedGenes[ geneName ][ repeatName ].append( round( ( commonPositions / geneLength ) * 100, 4 ) )\n",
    "    if geneName in affectedGenes :\n",
    "        if not \"Total\" in affectedGenes[ geneName ] :\n",
    "            affectedGenes[ geneName ][ \"Total\" ] = [ ]\n",
    "        affectedGenes[ geneName ][ \"Total\" ].append( min( 100.0, round( ( totalPositions / geneLength ) * 100, 4 ) ) )\n",
    "        \n",
    "affectedGeneNames = sorted( list( affectedGenes.keys( ) ) )\n",
    "relevantRepeatNames = list( relevantRepeats.keys( ) )\n",
    "relevantRepeatNames.sort( key = lambda rN : int( rN.split( \"_\" )[ 1 ] ) )\n",
    "relevantRepeatNamesANSorted = [ ]\n",
    "for rN in relevantRepeatNames :\n",
    "    if rN.startswith( \"NC\" ) :\n",
    "        relevantRepeatNamesANSorted.append( rN )\n",
    "for rN in relevantRepeatNames :\n",
    "    if rN.startswith( \"TP\" ) :\n",
    "        relevantRepeatNamesANSorted.append( rN )\n",
    "relevantRepeatNames = relevantRepeatNamesANSorted\n",
    "data2D = [ ]\n",
    "for relevantRepeatName in relevantRepeatNames :\n",
    "    data1D = [ ]\n",
    "    for affectedGeneName in affectedGeneNames :\n",
    "        if relevantRepeatName in affectedGenes[ affectedGeneName ] :\n",
    "            data1D.append( sum( affectedGenes[ affectedGeneName ][ relevantRepeatName ] ) )\n",
    "        else :\n",
    "            data1D.append( 0 )\n",
    "    data2D.append( data1D )\n",
    "total1D = [ ]\n",
    "for affectedGeneName in affectedGeneNames :\n",
    "    total1D.append( sum( affectedGenes[ affectedGeneName ][ \"Total\" ] ) )\n",
    "data2D.append( total1D )\n",
    "\n",
    "fig, ax = plt.subplots( figsize = ( 12, 10 ) )\n",
    "res = sns.heatmap( data2D,\n",
    "             cmap = \"GnBu\",\n",
    "             cbar_kws = {\n",
    "                 \"aspect\": 50,\n",
    "                 \"label\": \"Overlapping Positions (%)\",\n",
    "                 \"orientation\": \"horizontal\"\n",
    "             },\n",
    "             vmin = 0,\n",
    "             vmax = 100,\n",
    "             center = 50,\n",
    "             linewidths = 1,\n",
    "             annot = True,\n",
    "             annot_kws = {\n",
    "                 # 'rotation': 90,\n",
    "                 'va': 'center',\n",
    "                 'ha': 'center'\n",
    "             }\n",
    "           )\n",
    "for t in res.texts:\n",
    "    if t.get_text( ) == \"0\" :\n",
    "        t.set_text( \"\" )\n",
    "    elif t.get_text( ) == \"1e+02\" :\n",
    "        t.set_text( \"100 %\" )\n",
    "    else :\n",
    "        t.set_text( t.get_text( ) + \" %\" )\n",
    "ax.set_yticklabels( [ rN.split( \"-\" )[ 0 ].replace( \"NC021490.2\", \"\" ) + \" \" + rN.split( \"-\" )[ 1 ].replace( \"_\", \" \" ).replace( \"r\", \"R\" ) for rN in relevantRepeatNames ] + [ \"Total\" ], size = 12 )\n",
    "ax.set_ylabel( \"Repeats\", size = 18 )\n",
    "ax.set_xticklabels( affectedGeneNames, size = 14 )\n",
    "ax.set_xlabel( \"Genes\", size = 18 )\n",
    "plt.setp( ax.get_xticklabels( ), rotation = 55, ha = \"right\", rotation_mode = \"anchor\")\n",
    "plt.setp( ax.get_yticklabels( ), rotation = 0, ha = \"right\", rotation_mode = \"anchor\")\n",
    "ax.set_title( \"Overlapping Gene and Repeat Positions\", size = 20 )\n",
    "ax.figure.axes[-1].xaxis.label.set_size( 18 )\n",
    "ax.figure.axes[-1].tick_params( labelsize = 14 )\n",
    "fig.tight_layout( )\n",
    "plt.show( )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "educational-eating",
   "metadata": {},
   "source": [
    "### 3.5 Assess Disregarded Repeat Positions per Gene per Sample (without RecoverMI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "informed-range",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First assess the overlapping positions per gene and per repeat.\n",
    "perGeneRepeatPositions = { }\n",
    "for geneName in geneNames :\n",
    "    geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "    genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "    geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "    for repeatName, repeatInfo in repeats.items( ) :\n",
    "        commonPositions = set( )\n",
    "        for repeatStart in repeatInfo[ 'starts' ] :\n",
    "            repeatPositions = list( range( repeatStart, repeatStart + repeatInfo[ 'length' ] - 1 ) )\n",
    "            for gp in genePositions :\n",
    "                for rp in repeatPositions :\n",
    "                    if gp == rp :\n",
    "                        commonPositions.add( gp )\n",
    "        if len( commonPositions ) != 0 :\n",
    "            perGeneRepeatPositions[ geneName ] = commonPositions\n",
    "            \n",
    "# For each sample, assess the number of uncalled positions per gene and per repeat (for untreated .vcf).\n",
    "perSamplePerGeneUncalledCountUntreated = { }\n",
    "for sampleName in os.listdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling/\" ) :\n",
    "    perSamplePerGeneUncalledCountUntreated[ sampleName ] = { }\n",
    "    with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling/\" + sampleName + \"/\" + sampleName + \"-NC_021490.2-UG-variants.vcf\", \"r\" ) as untreatedCalls :\n",
    "        line = untreatedCalls.readline( )\n",
    "        while line :\n",
    "            if not line.startswith( \"#\" ) :\n",
    "                position = int( line.split( \"\\t\" )[ 1 ] )\n",
    "                for geneName in geneNames :\n",
    "                    if geneName in perGeneRepeatPositions.keys( ) and not geneName in perSamplePerGeneUncalledCountUntreated[ sampleName ] :\n",
    "                        perSamplePerGeneUncalledCountUntreated[ sampleName ][ geneName ] = 0\n",
    "                    if geneName in perGeneRepeatPositions.keys( ) and position in perGeneRepeatPositions[ geneName ] and line.split( \"\\t\" )[ 9 ].startswith( \"./.\" ) :\n",
    "                        perSamplePerGeneUncalledCountUntreated[ sampleName ][ geneName ] += 1\n",
    "            line = untreatedCalls.readline( )\n",
    "            \n",
    "from matplotlib.ticker import MultipleLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "republican-customs",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the percentage of uncalled positions per gene and per repeat and plot (for untreated .vcf).\n",
    "fig, axs = plt.subplots( 4, 3, figsize = ( 11, 15 ), sharex = True, sharey = True )\n",
    "axIdx0 = -1\n",
    "axIdx1 = 0\n",
    "for geneName in sorted( list( list( perSamplePerGeneUncalledCountUntreated.values( ) )[ 0 ].keys( ) ) ) :\n",
    "    axIdx0 += 1\n",
    "    if axIdx0 > 3 :\n",
    "        axIdx0 = 0\n",
    "        axIdx1 += 1\n",
    "    percentages = [ ]\n",
    "    ax = axs[ axIdx0 ][ axIdx1 ]\n",
    "    for sampleName in sorted( list( perSamplePerGeneUncalledCountUntreated.keys( ) ) ) :\n",
    "        geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "        genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "        geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "        percentages.append(\n",
    "            round( 100 * ( perSamplePerGeneUncalledCountUntreated[ sampleName ][ geneName ] / geneLength ), 4 )\n",
    "        )\n",
    "    ax.hist( percentages, bins = [ v + 1*10**-6 for v in np.arange( -5, 105, 5 ) ], color = \"dimgray\", rwidth = 0.9 )\n",
    "    ax.grid( )\n",
    "    ax.set_title( geneName, size = 18 )\n",
    "    if axIdx0 == 3 :\n",
    "        ax.set_xticks( np.arange( 0, 120, 20 ) )\n",
    "        ax.xaxis.set_tick_params( labelsize = 14 )\n",
    "        ax.set_xlabel( \"Disregarded Positions [%]\", size = 18 )\n",
    "    ax.xaxis.set_minor_locator(MultipleLocator(5))\n",
    "    if axIdx1 == 0 :\n",
    "        #ax.set_yticks( np.arange( 0.5, 74.5 ) )\n",
    "        ax.yaxis.set_tick_params( labelsize = 14 )\n",
    "        ax.set_ylabel( \"Count\", size = 18 )\n",
    "plt.suptitle( \"Histograms of Disregarded Repeat Positions in 74 Samples per Gene\\nRecoverMI not applied\", y = 1.01, size = 20 )\n",
    "plt.tight_layout( )\n",
    "plt.show( )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "colonial-choir",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the percentage of uncalled positions per gene and per repeat and plot (for untreated .vcf).\n",
    "data2D = [ ]\n",
    "for sampleName in sorted( list( perSamplePerGeneUncalledCountUntreated.keys( ) ) ) :\n",
    "    data1D = [ ]\n",
    "    for geneName in sorted( list( list( perSamplePerGeneUncalledCountUntreated.values( ) )[ 0 ].keys( ) ) ) :\n",
    "        geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "        genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "        geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "        data1D.append(\n",
    "            round( 100 * ( perSamplePerGeneUncalledCountUntreated[ sampleName ][ geneName ] / geneLength ), 4 )\n",
    "        )\n",
    "    data2D.append( data1D )\n",
    "\n",
    "fig, ax = plt.subplots( figsize = ( 12, 18 ) )\n",
    "res = sns.heatmap( data2D,\n",
    "     cmap = \"GnBu\",\n",
    "     cbar_kws = {\n",
    "         \"aspect\": 50,\n",
    "         \"label\": \"Disregarded Positions (%)\",\n",
    "         \"orientation\": \"horizontal\",\n",
    "         \"pad\": 0.08\n",
    "     },\n",
    "     vmin = 0,\n",
    "     vmax = 100,\n",
    "     center = 50,\n",
    "     linewidths = 1,\n",
    "     annot = True,\n",
    "     annot_kws = {\n",
    "         'va': 'center',\n",
    "         'ha': 'center'\n",
    "     }\n",
    ")\n",
    "for t in res.texts:\n",
    "    if t.get_text( ) == \"0\" :\n",
    "        t.set_text( \"\" )\n",
    "    elif t.get_text( ) == \"1e+02\" :\n",
    "        t.set_text( \"100 %\" )\n",
    "    else :\n",
    "        t.set_text( t.get_text( ) + \" %\" )\n",
    "ax.set_xticklabels( sorted( list( list( perSamplePerGeneUncalledCountUntreated.values( ) )[ 0 ].keys( ) ) ), size = 14 )\n",
    "ax.set_xlabel( \"Genes\", size = 18 )\n",
    "ax.set_yticks( np.arange( 0.5, 74.5 ) )\n",
    "ax.set_yticklabels( sorted( list( perSamplePerGeneUncalledCountUntreated.keys( ) ) ), size = 12 )\n",
    "ax.set_ylabel( \"Samples\", size = 18 )\n",
    "plt.setp( ax.get_xticklabels( ), rotation = 55, ha = \"right\", rotation_mode = \"anchor\")\n",
    "plt.setp( ax.get_yticklabels( ), rotation = 0, ha = \"right\", rotation_mode = \"anchor\")\n",
    "ax.set_title( \"Percentage of Disregarded Repeat Positions per Sample and Gene\\nRecoverMI not applied\", size = 20 )\n",
    "ax.figure.axes[-1].xaxis.label.set_size( 18 )\n",
    "ax.figure.axes[-1].tick_params( labelsize = 14 )\n",
    "fig.tight_layout( )\n",
    "plt.show( )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "closing-design",
   "metadata": {},
   "source": [
    "### 3.6 Assess Uncalled Repeat Positions per Gene per Sample (with RecoverMI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "narrow-payment",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each sample, assess the number of uncalled positions per gene and per repeat (for untreated .vcf).\n",
    "perSamplePerGeneUncalledCountTreated = { }\n",
    "for sampleName in os.listdir( \"./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling/\" ) :\n",
    "    perSamplePerGeneUncalledCountTreated[ sampleName ] = { }\n",
    "    with open( \"./R4_TPOMPeome_Hackl2022_VariantCalling/VariantCalling/\" + sampleName + \"/\" + sampleName + \"-NC_021490.2-UG-variants.fxd.vcf\", \"r\" ) as treatedCalles :\n",
    "        line = treatedCalles.readline( )\n",
    "        while line :\n",
    "            if not line.startswith( \"#\" ) :\n",
    "                position = int( line.split( \"\\t\" )[ 1 ] )\n",
    "                for geneName in geneNames :\n",
    "                    if geneName in perGeneRepeatPositions.keys( ) and not geneName in perSamplePerGeneUncalledCountTreated[ sampleName ] :\n",
    "                        perSamplePerGeneUncalledCountTreated[ sampleName ][ geneName ] = 0\n",
    "                    if geneName in perGeneRepeatPositions.keys( ) and position in perGeneRepeatPositions[ geneName ] and line.split( \"\\t\" )[ 9 ].startswith( \"./.\" ) :\n",
    "                        perSamplePerGeneUncalledCountTreated[ sampleName ][ geneName ] += 1\n",
    "            line = treatedCalles.readline( )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b6cdda",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.ticker import MultipleLocator\n",
    "\n",
    "# Compute the percentage of uncalled positions per gene and per repeat and plot (for untreated .vcf).\n",
    "fig, axs = plt.subplots( 4, 3, figsize = ( 11, 15 ), sharex = True, sharey = True )\n",
    "axIdx0 = -1\n",
    "axIdx1 = 0\n",
    "for geneName in sorted( list( list( perSamplePerGeneUncalledCountTreated.values( ) )[ 0 ].keys( ) ) ) :\n",
    "    axIdx0 += 1\n",
    "    if axIdx0 > 3 :  \n",
    "        axIdx0 = 0\n",
    "        axIdx1 += 1\n",
    "    percentages = [ ]\n",
    "    ax = axs[ axIdx0 ][ axIdx1 ]\n",
    "    for sampleName in sorted( list( perSamplePerGeneUncalledCountUntreated.keys( ) ) ) :\n",
    "        geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "        genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "        geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "        percentages.append(\n",
    "            round( 100 * ( perSamplePerGeneUncalledCountTreated[ sampleName ][ geneName ] / geneLength ), 4 )\n",
    "        )\n",
    "    ax.hist( percentages, bins = [ v + 1*10**-6 for v in np.arange( -5, 105, 5 ) ], color = \"dimgray\", rwidth = 0.9 )\n",
    "    ax.grid( )\n",
    "    ax.set_title( geneName, size = 18 )\n",
    "    if axIdx0 == 3 :\n",
    "        ax.set_xticks( np.arange( 0, 120, 20 ) )\n",
    "        ax.xaxis.set_tick_params( labelsize = 14 )\n",
    "        ax.set_xlabel( \"Disregarded Positions [%]\", size = 18 )\n",
    "    ax.xaxis.set_minor_locator(MultipleLocator(5))\n",
    "    if axIdx1 == 0 :\n",
    "        #ax.set_yticks( np.arange( 0.5, 74.5 ) )\n",
    "        ax.yaxis.set_tick_params( labelsize = 14 )\n",
    "        ax.set_ylabel( \"Count\", size = 18 )\n",
    "plt.suptitle( \"Histograms of Disregarded Repeat Positions in 74 Samples per Gene\\nRecoverMI applied\", y = 1.01, size = 20 )\n",
    "plt.tight_layout( )\n",
    "plt.show( )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "behavioral-burlington",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the percentage of uncalled positions per gene and per repeat and plot (for untreated .vcf).\n",
    "sampleNames = set( )\n",
    "geneNames = set( )\n",
    "data2D = [ ]\n",
    "for sampleName in sorted( list( perSamplePerGeneUncalledCountTreated.keys( ) ) ) :\n",
    "    data1D = [ ]\n",
    "    sampleNames.add( sampleName )\n",
    "    for geneName in sorted( list( list( perSamplePerGeneUncalledCountTreated.values( ) )[ 0 ].keys( ) ) ) :\n",
    "        geneInfo = annotatedGenes[ \"TPANIC_\" + geneName[ 2: ] ]\n",
    "        genePositions = list( range( geneInfo[ 'locusStart' ], geneInfo[ 'locusEnd' ] + 1 ) )\n",
    "        geneLength = geneInfo[ 'locusEnd' ] - geneInfo[ 'locusStart' ] + 1\n",
    "        data1D.append(\n",
    "            round( 100 * ( perSamplePerGeneUncalledCountTreated[ sampleName ][ geneName ] / geneLength ), 4 )\n",
    "        )\n",
    "        geneNames.add( geneName )\n",
    "    data2D.append( data1D )\n",
    "\n",
    "fig, ax = plt.subplots( figsize = ( 12, 18 ) )\n",
    "res = sns.heatmap( data2D,\n",
    "     cmap = \"GnBu\",\n",
    "     cbar_kws = {\n",
    "         \"aspect\": 50,\n",
    "         \"label\": \"Disregarded Positions (%)\",\n",
    "         \"orientation\": \"horizontal\",\n",
    "         \"pad\": 0.08\n",
    "     },\n",
    "     vmin = 0,\n",
    "     vmax = 100,\n",
    "     center = 50,\n",
    "     linewidths = 1,\n",
    "     annot = True,\n",
    "     annot_kws = {\n",
    "         'va': 'center',\n",
    "         'ha': 'center'\n",
    "     }\n",
    ")\n",
    "for t in res.texts:\n",
    "    if t.get_text( ) == \"0\" :\n",
    "        t.set_text( \"\" )\n",
    "    elif t.get_text( ) == \"1e+02\" :\n",
    "        t.set_text( \"100 %\" )\n",
    "    else :\n",
    "        t.set_text( t.get_text( ) + \" %\" )\n",
    "ax.set_xticklabels( sorted( list( list( perSamplePerGeneUncalledCountTreated.values( ) )[ 0 ].keys( ) ) ), size = 14 )\n",
    "ax.set_xlabel( \"Genes\", size = 18 )\n",
    "ax.set_yticks( np.arange( 0.5, 74.5 ) )\n",
    "ax.set_yticklabels( sorted( list( perSamplePerGeneUncalledCountTreated.keys( ) ) ), size = 12 )\n",
    "ax.set_ylabel( \"Samples\", size = 18 )\n",
    "plt.setp( ax.get_xticklabels( ), rotation = 55, ha = \"right\", rotation_mode = \"anchor\")\n",
    "plt.setp( ax.get_yticklabels( ), rotation = 0, ha = \"right\", rotation_mode = \"anchor\")\n",
    "ax.set_title( \"Percentage of Disregarded Repeat Positions per Sample and Gene\\nRecoverMI applied\", size = 20 )\n",
    "ax.figure.axes[-1].xaxis.label.set_size( 18 )\n",
    "ax.figure.axes[-1].tick_params( labelsize = 14 )\n",
    "fig.tight_layout( )\n",
    "plt.show( )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
